x-pack/plugins/endpoint/server/routes/resolver.ts

```diff
@@ -0,0 +1,181 @@
+ */
+
+import {
+  IRouter,
+  ResponseError,
+  Logger,
+  RequestHandler,
+  IScopedClusterClient,
+  KibanaResponseFactory,
+  IKibanaResponse,
+} from 'kibana/server';
+import { SearchResponse } from 'elasticsearch';
+import { schema } from '@kbn/config-schema';
+
+import { EndpointAppContext, Total, JSONish } from '../types';
+import {
+  getESChildrenQuery,
+  getESNodeQuery,
+  getESChildrenCountQuery,
+  getESNodeCountQuery,
+  PaginationInfo,
+} from '../services/resolver/query_builder';
+import { BaseSearchHandler } from '../services/resolver/search_handler';
+import { ResolverData } from '../../common/types';
+import { EntityParseError, CountQueryInfo } from '../services/resolver/common';
+import { ChildrenSearchHandler } from '../services/resolver/children_search_handler';
+import { SingleNodeHandler } from '../services/resolver/single_node_handler';
+
+interface ResolverRequestParams {
+  page_index?: number;
+  page_size?: number;
+  entity_id: string;
+}
+
+type SearchQueryBuilder = (
+  endpointContext: EndpointAppContext,
+  entityID: string,
+  paginationInfo: PaginationInfo
+) => Promise<JSONish>;
+
+type CountQueryBuilder = (entityID: string) => CountQueryInfo;
+
+type SearchHandlerCreator = (
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) => BaseSearchHandler;
+
+function createChildrenSearchHandler(
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) {
+  return new ChildrenSearchHandler(client, endpointContext, pageInfo, countQuery, entityID);
+}
+
+function createSingleNodeHandler(
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) {
+  return new SingleNodeHandler(client, endpointContext, pageInfo, countQuery, entityID);
+}
+
+function getReqHandler(
+  searchQueryBuilder: SearchQueryBuilder,
+  countQueryBuilder: CountQueryBuilder,
+  debugQueryString: string,
+  notFoundString: string,
+  log: Logger,
+  endpointAppContext: EndpointAppContext,
+  handlerCreator: SearchHandlerCreator
+): RequestHandler<unknown, ResolverRequestParams> {
+  return async (context, req, res) => {
+    const entityID = req.query.entity_id;
+    log.debug(`entity_id: ${entityID}`);
+    const paginationInfo = {
+      page: req.query?.page_index,
+      pageSize: req.query?.page_size,
+    } as PaginationInfo;
+    try {
+      const query = await searchQueryBuilder(endpointAppContext, entityID, paginationInfo);
+      log.debug(`${debugQueryString} ${JSON.stringify(query)}`);
+      const response = (await context.core.elasticsearch.dataClient.callAsCurrentUser(
+        'search',
+        query
+      )) as SearchResponse<ResolverData>;
+
+      if (response.hits.hits.length === 0) {
+        return res.notFound({ body: `${notFoundString}` });
```

Comment:
We should consider reserving 404's for meaning 'resource not found', which I would interpret as 'there is no such thing as resolver'. I think in this case we should just return empty collections and zeros for counts.